core/src/main/java/tc/oc/pgm/shops/menu/ShopMenu.java

```diff
@@ -142,35 +144,53 @@ private void setCategory(Category category, int slot) {
     this.highlight = slot;
   }
 
-  private ClickableItem[] getCategoryItems() {
-    ClickableItem[] items = new ClickableItem[shop.getVisibleCategories(getViewer()).size()];
-    for (int i = 0; i < shop.getCategories().size(); i++) {
-      items[i] = getCategoryItem(shop.getCategories().get(i));
+  private boolean updateCategories() {
+    boolean anyChanged = false;
+    for (CategoryItem category : allCategories) {
+      if (category.changed()) anyChanged = true;
+    }
+    if (categories != null && !anyChanged) return false;
```

Comment:
nit: Assuming we don't need to run `category.changed()` for every category every time:
```suggestion
    if (categories != null && allCategories.stream().noneMatch((CategoryItem::changed)))
      return false;
```

Reply:
We do need to check, since that will update the `active` flag on the category